### DISCUSSION TITLE
Question about identities

### DISCUSSION THREAD

**Discussion by murshex** [ravendb/ravendb#15684] [Q&A] [answered]
Hello,

I need to use identities (i.e. "accounts|") for account registration (email/password). I'm using a cluster wide transaction. It looks like identities are generated on the server when SaveChanges is called. Therefore I can't store an **email reference document** with the account id and create the **account document** all in one transaction.

What I would like to be able to do is to be able to get the account id (incrementing primary key) when Store is called, and if the transaction fails, I don't want the identity to be incremented.

Heres an example of my code.

```cs
// emails/test@example.com
public class EmailReference
{
    public string Id { get; set; }
    public string AccountId { get; set; }
}

// accounts/1
public class Account
{
    public string Id { get; set; }
}

public async Task Register(string email)
{
    using var session = DocumentStoreHolder.Store.OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide });

    var emailId = "emails/" + email;

    var emailRefDocument = await session.LoadAsync<EmailReference>(emailId);
    if (emailRefDocument != null)
        throw new ConcurrencyException();

    var account = new Account { Id = "accounts|" };

    await session.StoreAsync(account);
    // Problem: can't get account.Id because SaveChanges hasn't been called yet
    await session.StoreAsync(new EmailReference { AccountId = account.Id }, emailId);

    await session.SaveChangesAsync();
}
```

I want to use one transaction, to reserve the email (email reference document), create the account using identities for the account id, and if the transaction fails, I don't want to increment the identity.

Any suggestions?

**Comment by murshex** [comment, marked as answer]
https://groups.google.com/g/ravendb/c/HK67Pz9mm3Y

Looks like I asked this question before lol... !

> What you can do is something like this:
> Two compare exchanges, one that you'll increment to get the id and the other to reverse the value. Something like this:
> 
> ```cs
> var identity = session.Advanced.ClusterTransaction.GetCompareExchangeValue<long>("Identity");
> if (identity == null)
> {
>     identity = session.Advanced.ClusterTransaction.CreateCompareExchangeValue("Identity", 0L);
> }
> 
> identity.Value++;
> session.Advanced.ClusterTransaction.CreateCompareExchangeValue("users/ayende", identity.Value);
> session.SaveChanges();
> ```

If I do this, there is a chance that SaveChanges will fail because the Identity value changed in the meantime, right?

**Reply by georgiosd** [reply]
Take a step back - why do you *need* to use identity for accounts?

**Reply by murshex** [reply]
> Take a step back - why do you _need_ to use identity for accounts?

> In any other scenario, I would just use ravendb hilo document IDs. I cant do that in this scenario because Im trying to recreate a database for a video game that was originally using MySQL. If I could do it this way it would be a lot easier for me.
> Because of poor design. It's not something I can change.

**Reply by ayende** [reply]
Yes, in this scenario, the cluster wide tx would fail because of the changes to the identity value.

**Reply by georgiosd** [reply]
@c0nd3v the migration doesn't need to interfere with the design on the new system. Migrate the data over with the old ids then use Raven's default mechanism for the new data. That would work, wouldn't it?
